File: BOJ/2470.py
# 2470번: 두 용액
# 투 포인터

N = int(input())
values = list(map(int, input().split()))

values.sort()

i = 0
j = N - 1
min_sum = float('inf')
answer = (0, 0)

while i < j:
    temp_sum = values[i] + values[j]
    
    if abs(temp_sum) < min_sum:
        min_sum = abs(temp_sum)
        answer = (values[i], values[j])
        
    if temp_sum < 0:
        i += 1
    elif temp_sum > 0:
        j -= 1
    elif temp_sum == 0:
        break
        
print(answer[0], answer[1])

File: BOJ/2470.java
import java.util.*;
import java.io.*;

public class Main {

	public static void main(String[] args) throws Exception {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int N = Integer.parseInt(br.readLine());

		StringTokenizer st = new StringTokenizer(br.readLine());
		int[] arr = new int[N];
		for (int i = 0; i < N; i++) {
			arr[i] = Integer.parseInt(st.nextToken());
		}

		Arrays.sort(arr);

		int min_sum = Integer.MAX_VALUE;
		int s = 0;
		int e = N - 1;
		int[] answer = { 0, 0 };

		while (s < e) {
			int sum = arr[s] + arr[e];

			if (Math.abs(sum) < min_sum) {
				min_sum = Math.abs(sum);
				answer[0] = arr[s];
				answer[1] = arr[e];
			}

			if (sum == 0) {
				break;
			} else if (sum > 0) {
				e--;
			} else if (sum < 0) {
				s++;
			}

		}

		System.out.print(answer[0] + " ");
		System.out.print(answer[1] + "\n");

	}
}
